lms/models/event.py

```diff
@@ -0,0 +1,113 @@
+from enum import Enum
+
+import sqlalchemy as sa
+from sqlalchemy.dialects.postgresql import JSONB
+from sqlalchemy.ext.mutable import MutableDict
+
+from lms.db import BASE, varchar_enum
+
+
+class EventType(BASE):
+    __tablename__ = "event_type"
+
+    class Type(str, Enum):
+        CONFIGURED_LAUNCH = "configured_launch"
+        DEEP_LINKING = "deep_linking"
+
+    id = sa.Column(sa.Integer, autoincrement=True, primary_key=True)
+    type = varchar_enum(Type)
+
+
+class Event(BASE):
+    """Model to store any relevant events that occur within the application."""
+
+    __tablename__ = "event"
+
+    id = sa.Column(sa.Integer, autoincrement=True, primary_key=True)
+
+    timestamp = sa.Column(
+        sa.DateTime(), server_default=sa.func.now(), nullable=False, index=True
+    )
+    """Time the event occurred, defaults to now() if not specified"""
+
+    type_id = sa.Column(
+        sa.Integer(), sa.ForeignKey("event_type.id", ondelete="cascade")
+    )
+    type = sa.orm.relationship("EventType")
+
+    """One of EventType"""
+
+    application_instance_id = sa.Column(
+        sa.Integer(),
+        sa.ForeignKey("application_instances.id", ondelete="cascade"),
+        nullable=True,
+    )
+    application_instance = sa.orm.relationship("ApplicationInstance")
+    """Which application instance this events relates to"""
+
+    course_id = sa.Column(
+        sa.Integer(), sa.ForeignKey("grouping.id", ondelete="cascade"), nullable=True
+    )
+    course = sa.orm.relationship("Course", foreign_keys=[course_id])
+    """If this event relates to one course only, which one"""
+
+    assignment_id = sa.Column(
+        sa.Integer(), sa.ForeignKey("assignment.id", ondelete="cascade"), nullable=True
+    )
+    assignment = sa.orm.relationship("Assignment")
+    """Which assignment this event relates to"""
+
+    grouping_id = sa.Column(
+        sa.Integer(), sa.ForeignKey("grouping.id", ondelete="cascade"), nullable=True
+    )
+    grouping = sa.orm.relationship("Grouping", foreign_keys=[grouping_id])
+    """If this event belongs to one particular grouping, which one"""
+
```

Comment:
A relationship to the data and users would be very helpful here. Although as we discussed, hard to get right without any calling code, so perhaps now isn't the time.